src/clp_ffi_js/ir/ClpIrV1Decoder.cpp

```diff
@@ -0,0 +1,233 @@
+    }
+
+    auto results{emscripten::val::object()};
+    results.set("numValidEvents", m_log_events.size());
+    results.set("numInvalidEvents", 0);
+    return results;
+}
+
+auto ClpIrV1Decoder::decode(size_t begin_idx, size_t end_idx) const -> emscripten::val {
+    if (m_log_events.size() < end_idx || begin_idx >= end_idx) {
+        return emscripten::val::null();
+    }
+
+    size_t log_num{begin_idx + 1};
+    std::string message;
+    constexpr size_t cDefaultReservedMessageLength{512};
+    message.reserve(cDefaultReservedMessageLength);
+    auto const results{emscripten::val::array()};
+    std::span<clp::ir::LogEvent<clp::ir::four_byte_encoded_variable_t> const> const log_events_span(
+            m_log_events.begin() + static_cast<std::ptrdiff_t>(begin_idx),
+            m_log_events.begin() + static_cast<std::ptrdiff_t>(end_idx)
+    );
+    for (auto const& log_event : log_events_span) {
+        message.clear();
+
+        // TODO: Replace below handlers code by an OSS decoding method once it's added in the future
+        auto const constant_handler = [&](std::string const& value, size_t begin_pos, size_t length
+                                      ) { message.append(value, begin_pos, length); };
+
+        auto const encoded_int_handler = [&](clp::ir::four_byte_encoded_variable_t value) {
+            message.append(clp::ffi::decode_integer_var(value));
+        };
+
+        auto const encoded_float_handler
+                = [&](clp::ir::four_byte_encoded_variable_t encoded_float) {
+                      message.append(clp::ffi::decode_float_var(encoded_float));
+                  };
+
+        auto const dict_var_handler
+                = [&](std::string const& dict_var) { message.append(dict_var); };
+
+        try {
+            clp::ffi::ir_stream::generic_decode_message<true>(
+                    log_event.get_logtype(),
+                    log_event.get_encoded_vars(),
+                    log_event.get_dict_vars(),
+                    constant_handler,
+                    encoded_int_handler,
+                    encoded_float_handler,
+                    dict_var_handler
+            );
+        } catch (clp::ffi::ir_stream::DecodingException const& e) {
+            SPDLOG_ERROR("Failed to decode. Error Code: {}", e.get_error_code());
+            break;
+        }
+
+        constexpr size_t cLogLevelNone{0};
+        constexpr size_t cLogLevelBeginIdx{cLogLevelNone + 1};
+        constexpr size_t cLogLevelPositionInMessages{1};
+        size_t log_level{cLogLevelNone};
+        for (size_t i{cLogLevelBeginIdx}; i < cLogLevelNames.size(); ++i) {
+            if (message.substr(cLogLevelPositionInMessages).starts_with(cLogLevelNames[i])) {
+                log_level = i;
+                break;
+            }
+        }
+
+        m_ts_pattern.insert_formatted_timestamp(log_event.get_timestamp(), message);
+
+        EM_ASM(
+                { Emval.toValue($0).push([UTF8ToString($1), $2, $3, $4]); },
```

Comment:
We've seen users logging things that are not UTF-8, so shall we switch this to `AsciiToString`?

Reply:
Not sure if I understand the proposed solution. I thought ASCII is a subset of UTF-8 rather than the other way around, so switching this to `AsciiToString` would cause more incompatibilities?

e.g. I tried `AsciiToString("👋")` which gave `'ð\x9F\x91\x8B'`.

Reply:
Right, sorry, should've thought it through more and explained my reasoning.

What I was getting at is:

* Sometimes a user may compress a log with a byte (or bytes) that are invalid UTF-8 characters (e.g., `0xFF`).
* In Python, if you try to decode 0xff as a UTF-8 character (`b"\xFF".decode()`), it will error out because it's an invalid UTF-8 character. So in the past, we've take the approach of decoding it as a latin-1 character so that the stream reader doesn't error out.
 * This isn't true in clp-ffi-py though: if you call `get_log_message`, it will error out, whereas if you call `get_formatted_message`, it will format the invalid characters as '�'.
 * That said, when rendering the logs, '�' is probably better than a random latin-1 character. If the user needs the original byte(s), they should access the stream programmatically (and we should provide APIs to do so).
* It turns out `UTF8ToString` doesn't error out on invalid bytes and instead returns '�'.

So overall, yeah, we should stick with `UTF8ToString`; but, we should probably refine these APIs in future PRs. Two issues I see right now are:

* We don't provide an API to just return the log event without formatting it.
* The way we insert the timestamp into the formatted log event is not very efficient considering we *insert* it the beginning rather than starting the formatted string with the timestamp. This is an artifact of how `TimestampPattern` works and the use cases it needs to support (timestamps don't always appear first in an unstructured log event), but it's something we should address for IR streams since we expect the user to define the timestamp format they want to use rather than using something predefined.

Reply:
Gotcha. will keep those in mind